### ISSUE TITLE
O(N^2) performance in the number of "--bind" flags

### ISSUE THREAD

**Issue by brown** [containers/bubblewrap#384]
When bwrap is called with thousands of "--bind" arguments on the command line, it slows down considerably.  It looks like /proc/self/mountinfo is parsed for every single "--bind" flag.  Must that happen?  Can the code instead parse the mount table once and update its representation over time?

**Comment by brown** [comment]
Some more information ....

As bwrap processes each "--bind" flags it parses /proc/self/mountinfo.  After every bind mount that file is longer by one line ... that's what leads to N^2 behavior.

I ran a test where I pass bwrap 2000 "--bind" flags ...
    bwrap --dev-bind / / --bind /usr /tmp/1  --bind /usr /tmp/2 ... --bind /usr /tmp/2000 /bin/true

The program takes 6 or 7 seconds to run.  If I hack the code to re-use information collected from parsing /proc/self/mountinfo, the same command takes about 150 milliseconds to run.

**Comment by L-as** [comment]
You could open a PR, so your change could (presumably) be subject to review.

**Comment by brown** [comment]
Here is a pull request I cooked up to solve the problem.  It likely has some bugs.  I'd like to work with the bubblewrap maintainers to land it.  Thanks!.

**Comment by brown** [comment]
I have updated the pull request:  https://github.com/containers/bubblewrap/pull/385
I've done some testing, so it is now much more likely to be correct.

**Comment by alexlarsson** [comment]
It kinda does this intentionally, because its pretty hard to know exactly how a recursive bind operation affected the system, especially since other things could be happening in parallel (like e.g. mount propagation). It is particularly sensitive to get this right in the case where bubblewrap is setuid as getting this wrong might be a possible attack vector.

Not saying it is impossible, but it needs to be super careful.

**Comment by brown** [comment]
Is there a different strategy that might work to reduce the number of times
the mountinfo file is read?
I don't fully understand mount propagation and the need to adjust flags for
recursive mounts.
My use case is a build system, where bubblewrap is used to create a
sandbox.  It's not unusual
to bind mount 2K paths, which takes a really long time with the current
implementation.

**Comment by geofft** [comment]
cc @kolyshkin - do you have ideas on how to robustly maintain some in-memory idea of what mountinfo looks like without rereading mountinfo on each mount operation?

**Comment by kolyshkin** [comment]
Ahh, what a nice case.

I guess the problem here is 
 - every bind mount might result in multiple changes in mountinfo (depending on mount propagation and recursiveness);
 - the order of mounts is important.

So, to avoid re-reading the mountinfo, we have to replicate the kernel logic of propagating the mounts, and this seems like being too much. Currently, though, it's the only way to go.

The only other thing I can think of is trying to optimize the parser, and maybe finding some specific cases in which we can avoid re-reading mountinfo.

From the perspective of a user, 

> It's not unusual to bind mount 2K paths

it might make sense to figure out a way to decrease the number of bind mounts.

**Comment by geofft** [comment]
Something that might help here: in our use case, we know that each directory being bind-mounted has no nested mounts inside it, and we don't care about propagation to/from it (i.e., it would be fine to unconditionally mark them as private).

Could we make `bind_mount` detect whether something being mounted in has any current nested mounts, and avoid rereading mountinfo if so? That way we know we don't have to do any calculation on our own that affects things when doing a series of such mounts.

Or perhaps a separate option `--bind-nonrecursive` (which fails if there's anything inside it)?

**Comment by kolyshkin** [comment]
> Could we make bind_mount detect whether something being mounted in has any current nested mounts

funny, but such detection requires reading mountinfo, so it does not make sense (reading mountinfo to avoid reading mountinfo).

> Or perhaps a separate option --bind-nonrecursive (which fails if there's anything inside it)?

It seems like in case of non-recursive bind mount we can avoid reading mountinfo, as the only thing we need is flags and it can be obtained via statfs(2).

**Comment by geofft** [comment]
Yes, but you can do it _once_ - i.e. if you have `bwrap --bind /1 /1 ... --bind /2000 /2000`, you can read mountinfo once, determine that 1 through 2000 are all non-nested, and then bind-mount them all without rereading mountinfo.

Also I think you can detect it by trying the mount without MS_REC and seeing if it works, right? It will fail if the mount has other mounts inside it and you aren't privileged over the original mount namespace, so potentially we could do something like
```c
for (...) {
    if (mount(source, target, NULL, MS_BIND | MS_PRIVATE, NULL) == 0)
        continue;
    if (errno == EINVAL) {
        if (mount(source, target, NULL, MS_BIND | MS_REC | MS_PRIVATE, NULL) == 0) {
            reread_mountinfo();
            continue;
        }
    }
    return -1;
}
```

I believe at this point in the code, since we've made a new mount namespace, we know we're not privileged over the original one.

**Comment by haampie** [comment]
I was also surprised skimming through the code that every `bind_mount` call goes over mountinfo.

My use case is as follows: replace `--dev-bind / /` with `--dev-bind /bin /bin --dev-bind /usr /usr ...` for every dir entry of `/`, so that I can bind a new path `--bind ./hello-world /hello-world` at the root of my filesystem.

It'd be nice if there was a flag like `--dev-bind-children / / --bind ./hello-world /hello-world` or so, that does this efficiently

**Comment by witaway** [comment]
When I've seen this Issue, I became that much interested. I think I have one idea, don't know is it fit the problem or not.

Maybe we can handle `--bind`s not directly one by one doing extra work on each call rechecking `/proc/self/mountinfo`, but we could save all these "requests" in array and then optimize work knowing all information about problem?

For example, we could build a tree, where root is root and every other node is a part of path? Also we could add additional flag to each node: 
1. Is it mount point or just regular path of the path?
2. What's the mount root pointing to this mount point? If it's mountpoint.
3. List of children folders.

Next we iterate from first bind operation to last and see if it require `NODEV` or `RDONLY` flag. We set this flag to corresponding node and propagate change to all of the childen.

It's still O(N^2) if we propagate it naively.

But we can start DFS from root and assign:
1. `start[node]` - which denotes the time at which we enter this node during DFS
2. `finish[node` - which denotes the time at which we leave this node during DFS.

So-called "time" is an argument of DFS function and when we call DFS into children we give it time incremented by one.

Now for any node say `X` and another node `Y` in subtree, `start[X] < start[Y] <= finish[Y] <= finish[X]`. So when we set/unset `NODEV`/`RDONLY` flag to subtree of X, we do a lazy update of adding flag to all indices from `start[X]` to `finish[X]`.

And lazy propagation of flag on this interval we can do with a segment/fenwick tree. Segment is simpler and just enough.Segment tree is data structure that is useful when you have an array with values and you want to:
1. Change values of array fast.
2. Query sum on interval fast.

When we set flag on subtree, we assign 1 on all the needed interval.
When we unset flag on subtree, we assign 0 on interval.
When we check if flag set on somewhat node, we just query the sum of 1-length interval `[start[X], start[X]]` on segment tree.

There's already ton of implementations of segment tree with lazy assignment in the internet, so implementation is not much big problem.

It's like we need to create two identical segment trees: one for `NODEV` flags and one for `RDONLY` flag. But, now I'm not sure, I think we can optimize it that way:
1. Assign `0` (`00b`) - for no flags
2. Assign `1` (`01b`) - for `NODEV` flag
3. Assign `2` (`10b`) - for `RDONLY` flag
4. Assign `4` (`11b`) - for both

Then, when we can retrieve the final state of `NODEV/RDONLY` flags after all `--bind` operations, we just mount all mountpoints with already known flags. 

I think it might increase the speed dramatically. Segment tree is `O(log N)` - for both assign and query operations, and `O(4N)` in memory sense - not so much for any reasonable quantity of mountpoints.

What do you think?

**Comment by ambiso** [comment]
The new mount APIs would be a way to get rid of the performance issue. (using fsopen, fsconfig, fsconfig_create, fsmount, open_tree, move_mount, move_bind)

**Comment by smcv** [comment]
> The new mount APIs would be a way to get rid of the performance issue. (using fsopen, fsconfig, fsconfig_create, fsmount, open_tree, move_mount, move_bind)

If someone can implement that (with graceful fallback to the current way, to accommodate older kernels), that would be the right answer to this.

**Comment by smcv** [comment]
> The new mount APIs would be a way to get rid of the performance issue. (using fsopen, fsconfig, fsconfig_create, fsmount, open_tree, move_mount, move_bind)

To get rid of the performance issue, I think `mount_setattr()` is probably enough: I've opened #754 for that idea.

I don't think using the rest of the new mount APIs is necessarily required for the performance issue, but it could help us with making `--[ro-]bind-fd` more robust: #755.

**Comment by smcv** [comment]
In current git main (to be released as 0.13.0 in future), performance should only be O(n^2) in the number of mounts if you have a very old kernel that doesn't support `mount_setattr()`. I think that's probably good enough to consider this solved?
